File: Week_01/84/largest_rectangle_in_histogram3.go
package main

func largestRectangleArea(heights []int) int {
	res := 0
	var stack []int
	heights = append([]int{0}, heights...)
	heights = append(heights, 0)

	for i, h := range heights {
		for len(stack) > 1 && h < heights[stack[len(stack)-1]] {
			idx := stack[len(stack)-1]
			res = max(res, heights[idx]*(i-stack[len(stack)-2]-1))
			stack = stack[:len(stack)-1]
		}
		stack = append(stack, i)
	}
	return res
}

File: Week_01/84/largest_rectangle_in_histogram3.py
from typing import List


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        res = 0
        stack = []
        heights = [0] + heights + [0]

        for i, h in enumerate(heights):
            while len(stack) > 1 and h < heights[stack[-1]]:
                idx = stack.pop()
                res = max(res, heights[idx] * (i - stack[-1] - 1))
            stack.append(i)
        return res

    def test(self):
        for heights, target in [
            ([], 0),
            ([2, 1, 5, 6, 2, 3], 10),
        ]:
            ans = self.largestRectangleArea(heights)
            assert ans == target, f"target: {target}, ans {ans}"

        print("all done")


if __name__ == "__main__":
    Solution().test()
